ArduPlane/events.cpp

```diff
@@ -16,34 +16,34 @@ void Plane::failsafe_short_on_event(enum failsafe_state fstype, mode_reason_t re
     case FLY_BY_WIRE_B:
     case CRUISE:
     case TRAINING:
-        failsafe.saved_mode = control_mode;
+        failsafe.saved_mode = (FlightMode)control_mode->mode_number();
         failsafe.saved_mode_set = true;
         if(g.fs_action_short == FS_ACTION_SHORT_FBWA) {
-            set_mode(FLY_BY_WIRE_A, reason);
+            set_mode(mode_fbwa, reason);
         } else {
-            set_mode(CIRCLE, reason);
+            set_mode(mode_circle, reason);
         }
         break;
 
     case QSTABILIZE:
     case QLOITER:
     case QHOVER:
-        failsafe.saved_mode = control_mode;
+        failsafe.saved_mode = (FlightMode)control_mode->mode_number();
         failsafe.saved_mode_set = true;
-        set_mode(QLAND, reason);
+        set_mode(mode_qland, reason);
         break;
         
     case AUTO:
     case AVOID_ADSB:
     case GUIDED:
     case LOITER:
         if(g.fs_action_short != FS_ACTION_SHORT_BESTGUESS) {
-            failsafe.saved_mode = control_mode;
+            failsafe.saved_mode = (FlightMode)control_mode->mode_number();
             failsafe.saved_mode_set = true;
```

Comment:
Getting rid of this and making `failsafe.saved_mode` ought to be high on the to-do-next list.